lib/linkerd2-proxy-discover/src/buffer.rs

```diff
@@ -0,0 +1,161 @@
+use futures::{try_ready, Async, Future, Poll, Stream};
+use linkerd2_never::Never;
+use linkerd2_proxy_core::Error;
+use linkerd2_task as task;
+use std::fmt;
+use tokio::sync::{mpsc, oneshot};
+use tower::discover;
+use tracing::info_span;
+use tracing_futures::Instrument;
+
+#[derive(Clone, Debug)]
+pub struct Buffer<M> {
+    capacity: usize,
+    inner: M,
+}
+
+#[derive(Debug)]
+pub struct Discover<K, S> {
+    rx: mpsc::Receiver<discover::Change<K, S>>,
+    _disconnect_tx: oneshot::Sender<Never>,
+}
+
+pub struct DiscoverFuture<T, F, D> {
+    target: T,
+    future: F,
+    capacity: usize,
+    _marker: std::marker::PhantomData<fn() -> D>,
+}
+
+pub struct Daemon<D: discover::Discover> {
+    discover: D,
+    disconnect_rx: oneshot::Receiver<Never>,
+    tx: mpsc::Sender<discover::Change<D::Key, D::Service>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Lost(());
```

Comment:
It's a little confusing to me that we have both this error and the `ResolutionLost` error in `proxy-core`. If I were downcasting errors upstream of this, I think it might be unclear what type to check for. What about adding a `new` to `ResolutionLost` and re-using it here? Or, if these errors have different meanings, a comment and/or name change could be good.

Reply:
in one the resolution was lost because a stream completed, in the other, a task failed unexpectedly, so they are semantically different. ==

Reply:
> [..] they are semantically different.

in that case, i think we should probably rename them/change the `Display` impls/add comments to make that clearer...

Reply:
ftr: the blanket impl on Stream was removed, as it wasn't necessary